Refactor tests

jdudley1123 opened the pull request:
## Changes in this PR
This PR refactors lots of tests. It does not change the app's functionality.

Each commit is self contained, and generally covers one of:
- DRYing up tests, factories, and fixtures
- Enforcing consistent behaviour
- Removing unused/unnecessary code
- Refactoring to make code more readable and easier to maintain

jdudley1123 committed:
Refactor API forecasts fixture

Break the method into three separate methods that can be called individually for greater flexibility, and replace JSON with native Ruby objects for better readability.

jdudley1123 committed:
Refactor factory initialization functions

Refactory if you will...

Use shorter notation given that we just want all the attributes. Uses the double splat (`**`) operator to convert the attribute hash into keyword arguments for the class initializer.

jdudley1123 committed:
Remove unused factory traits

These traits are not used anywhere.

jdudley1123 committed:
Refactor feature test steps

Current we are previously using Cucumber-like method names and defining test steps in a separate file.

While Cucumber syntax makes understanding a scenario slightly easier, it makes it much harder to understand the code itself as we must flick between the spec and the steps. Moving the code back into the spec and adding comments gives us the best of both worlds.

The vast majority of the steps are not reusable, resulting in a lot of additional code for no benefit. Those steps that are reuseable, we move to a forecast helper.

jdudley1123 committed:
Initialize models in a consistent way

Other models like `forecast_zone` and `temperature_prediction` use keyword arguments, and we should be consistent.

jdudley1123 committed:
Use FactoryBot builder rather than initializing the class directly

Using FactoryBot means we don't have to worry about providing dummy values for attributes we don't care about.

jdudley1123 committed:
Move environmental variables to .env.test file

Defining environmental variables that are unrelated to the test is confusing. Variables that are required to get the app to run correctly should be moved to the `.env.test` file we already have to hand.

We use ClimateControl when we're actually testing the behaviour under different environmental conditions.

jdudley1123 committed:
Use ClimateControl to set environmental variables

Per ThoughtBot's guidance (https://thoughtbot.com/blog/testing-and-environment-variables#use-climate-control), it's better not to interact with ENV directly.

patrickjfl commented on spec/factories/forecasts.rb:
```diff
@@ -23,51 +23,16 @@
-    trait :air_pollution_level_1 do
-      air_pollution { FactoryBot.build(:air_pollution_prediction, 1) }
+    transient do
+      air_pollution_level { 1 }
```
nice, i was trying to work out how to do this!

patrickjfl commented on spec/factories/air_pollutions_predictions.rb:
```diff
@@ -53,96 +53,6 @@
-      particulate_matter_2_5 { 6 }
-      ozone { 6 }
-    end
-
-    trait 7 do
-      value { 7 }
-      label { "HIGH" }
-      nitrogen_dioxide { 7 }
-      particulate_matter_10 { 7 }
-      particulate_matter_2_5 { 7 }
-      ozone { 7 }
-    end
-
-    trait 8 do
-      value { 8 }
-      label { "HIGH" }
-      nitrogen_dioxide { 8 }
-      particulate_matter_10 { 8 }
-      particulate_matter_2_5 { 8 }
-      ozone { 8 }
-    end
-
-    trait 9 do
-      value { 9 }
-      label { "HIGH" }
-      nitrogen_dioxide { 9 }
-      particulate_matter_10 { 9 }
-      particulate_matter_2_5 { 9 }
-      ozone { 9 }
-    end
-
-    trait 10 do
-      value { 10 }
-      label { "VERY HIGH" }
-      nitrogen_dioxide { 10 }
-      particulate_matter_10 { 10 }
-      particulate_matter_2_5 { 10 }
-      ozone { 10 }
-    end
-
```
ah yes good spot, I forgot to remove these

patrickjfl commented on .env.test:
```diff
@@ -3,7 +3,12 @@
 
 DATABASE_URL=postgres://postgres@localhost:5432/air-text-test
 HOSTNAME=localhost
+
+CERC_API_HOST_URL=https://cerc.example.com,
+CERC_API_KEY=SECRET-API-KEY,
+CERC_API_CACHE_LIMIT_MINS=60,
+MAPTILER_API_KEY=TOPSECRET
```
makes sense 👍

patrickjfl commented on spec/features/visitors/view_forecasts_spec.rb:
```diff
@@ -1,58 +1,112 @@
+
+        # Predicted temperature level for tomorrow
+        expect_prediction(category: :temperature, level: :moderate)
+      end
```
I'm not sure about this one. IMO we are sacrificing quite a bit in terms of readability for what we gain here...

jdudley1123 replied:
I've addedd Cucumber-style comments to improve readability

patrickjfl commented on spec/components/day_tab_component_spec.rb:
```diff
@@ -6,7 +6,7 @@
     context "when the air pollution level is 1" do
       let(:component) {
-        DayTabComponent.new(forecast: FactoryBot.build(:forecast, air_pollution_level: 1), day: :today)
+        DayTabComponent.new(forecast: build(:forecast, air_pollution_level: 1), day: :today)
```
It may not be necessary, but is it clearer? Just thinking in terms of handing this over to CERC, who don't have Rails experience.

jdudley1123 replied:
Agreed, I've reverted this change, and made sure we're explicitly calling FactoryBot for all build and create calls.

patrickjfl commented on spec/models/forecast_zone_spec.rb:
```diff
@@ -1,7 +1,7 @@
-    let(:borough) { ForecastZone.new(type: 1, name: double, id: double) }
-    let(:area) { ForecastZone.new(type: 2, name: double, id: double) }
+    let(:borough) { build(:forecast_zone, type: 1) }
+    let(:area) { build(:forecast_zone, type: 2) }
```
nice

patrickjfl reviewed (approved):
Looks good to me, lots of clear improvements - just a couple that I'm less sure about.

jdudley1123 committed:
DRY up forecasts factory

Use a transient attribute to generate an air pollution prediction with an arbitrary level rather than having to define a separate trait for each one.
